src/cdk/table/table.spec.ts

```diff
@@ -523,24 +563,97 @@ describe('CdkTable', () => {
     });
 
     it('should error if there is row data that does not have a matching row template',
-      fakeAsync(() => {
-        expect(() => {
-          try {
-            createComponent(WhenRowWithoutDefaultCdkTableApp).detectChanges();
-            flush();
-          } catch {
-            flush();
-          }
-        }).toThrowError(getTableMissingMatchingRowDefError().message);
-    }));
-
-    it('should error if there are multiple rows that do not have a when function', fakeAsync(() => {
+        fakeAsync(() => {
+          const whenRowWithoutDefaultFixture = createComponent(WhenRowWithoutDefaultCdkTableApp);
+          const data = whenRowWithoutDefaultFixture.componentInstance.dataSource.data;
+          expect(() => {
+            try {
+              whenRowWithoutDefaultFixture.detectChanges();
+              flush();
+            } catch {
+              flush();
+            }
+          }).toThrowError(getTableMissingMatchingRowDefError(data[0]).message);
+        }));
+
+    it('should fail when multiple rows match data without multiTemplateDataRows', fakeAsync(() => {
       let whenFixture = createComponent(WhenRowMultipleDefaultsCdkTableApp);
       expect(() => {
         whenFixture.detectChanges();
         flush();
       }).toThrowError(getTableMultipleDefaultRowDefsError().message);
     }));
+
+    describe('with multiTemplateDataRows', () => {
+      it('should be able to render multiple rows per data object', () => {
+        setupTableTestApp(WhenRowCdkTableApp);
+        component.multiTemplateDataRows = true;
+        fixture.detectChanges();
+
+        const data = component.dataSource.data;
+        expectTableToMatchContent(tableElement, [
+          ['Column A', 'Column B', 'Column C'],
+          [data[0].a, data[0].b, data[0].c],
+          [data[1].a, data[1].b, data[1].c],
+          ['index_1_special_row'],
+          [data[2].a, data[2].b, data[2].c],
+          ['c3_special_row'],
+          [data[3].a, data[3].b, data[3].c],
+        ]);
+      });
+
+      it('should have the correct data and row indicies', () => {
+        setupTableTestApp(WhenRowCdkTableApp);
+        component.multiTemplateDataRows = true;
+        component.showIndexColumns();
+        fixture.detectChanges();
+
+        expectTableToMatchContent(tableElement, [
+          ['Index', 'Data Index', 'Render Index'],
+          ['', '0', '0'],
+          ['', '1', '1'],
+          ['', '1', '2'],
+          ['', '2', '3'],
+          ['', '2', '4'],
+          ['', '3', '5'],
+        ]);
+      });
+
+      it('should have the correct data and row indicies when data is homogeneous', () => {
```

Comment:
"Homogeneous" doesn't really communicate the same thing as "Data contains multiple occurrences of the same object instance"